cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -81,24 +89,58 @@
 
   private final TransactionExecutorFactory factory;
   private final ScheduleStoreTableUtil tableUtil;
+  private final AtomicBoolean appVerUpgrade;
+  private final AtomicBoolean storeInitialized;
+  private ExecutorService executorService;
   private Table table;
 
   @Inject
   public DatasetBasedStreamSizeScheduleStore(TransactionExecutorFactory factory, ScheduleStoreTableUtil tableUtil) {
     this.tableUtil = tableUtil;
     this.factory = factory;
+    this.appVerUpgrade = new AtomicBoolean(false);
+    this.storeInitialized = new AtomicBoolean(false);
   }
 
   /**
    * Initialize this persistent store.
    */
   public synchronized void initialize() throws IOException, DatasetManagementException {
+    initializeTable();
+
+    // Check whether the app upgrade is complete, if it is, then don't have to start a upgrade checker thread
+    checkUpgradeComplete(appVerUpgrade);
+    if (!appVerUpgrade.get()) {
+      executorService = Executors.newSingleThreadExecutor(
+        Threads.createDaemonThreadFactory("streamsize-schedulestore-appver-upgrade-checker"));
```

Comment:
If we use a LoadingCache with a timeout to check for upgrade column in HBase table in class `UpgradeCompleteCheck`, we don't need to start upgrade checker threads. 

Until the upgrade is complete,  the loading cache will check the HBase table every timeout time to see if upgrade is complete. All other times it will return the cached value. Once the upgrade is done, a flag is set and the loading cache is no longer consulted.